src/main.cpp

```diff
@@ -2265,209 +2163,287 @@ void MainFrame::startRxStream()
 
-        // Start sound card 1 ----------------------------------------------------------
-
-        m_rxInPa->setUserData(g_rxUserdata);
-        m_rxErr = m_rxInPa->setCallback(rxCallback);
-
-        m_rxErr = m_rxInPa->streamOpen();
-
-        if(m_rxErr != paNoError) {
-            wxMessageBox(wxT("Sound Card 1 Open/Setup error."), wxT("Error"), wxOK);
-			delete m_rxInPa;
-			if(two_rx)
-				delete m_rxOutPa;
-			delete m_txInPa;
-			if(two_tx)
-				delete m_txOutPa;
-            destroy_fifos();
-            destroy_src();
-            deleteEQFilters(g_rxUserdata);
-            delete g_rxUserdata;
-            m_RxRunning = false;
-            Pa_Terminate();
-            return;
-        }
-
-        m_rxErr = m_rxInPa->streamStart();
-        if(m_rxErr != paNoError) {
-            wxMessageBox(wxT("Sound Card 1 Stream Start Error."), wxT("Error"), wxOK);
-			delete m_rxInPa;
-			if(two_rx)
-				delete m_rxOutPa;
-			delete m_txInPa;
-			if(two_tx)
-				delete m_txOutPa;
-            destroy_fifos();
-            destroy_src();
-            deleteEQFilters(g_rxUserdata);
-            delete g_rxUserdata;
-            m_RxRunning = false;
-            Pa_Terminate();
-            return;
-        }
-
-        // Start separate output stream if needed
-
-        if(two_rx) {
-            m_rxOutPa->setUserData(g_rxUserdata);
-            m_rxErr = m_rxOutPa->setCallback(rxCallback);
-
-            m_rxErr = m_rxOutPa->streamOpen();
-
-            if(m_rxErr != paNoError) {
-                wxMessageBox(wxT("Sound Card 1 Second Stream Open/Setup error."), wxT("Error"), wxOK);
-                delete m_rxInPa;
-                delete m_rxOutPa;
-                if(two_tx)
-                    delete m_txOutPa;
-                destroy_fifos();
-                destroy_src();
-                deleteEQFilters(g_rxUserdata);
-                delete g_rxUserdata;
-                m_RxRunning = false;
-                Pa_Terminate();
-                return;
-            }
-
-            m_rxErr = m_rxOutPa->streamStart();
-            if(m_rxErr != paNoError) {
-                wxMessageBox(wxT("Sound Card 1 Second Stream Start Error."), wxT("Error"), wxOK);
-                m_rxInPa->stop();
-                m_rxInPa->streamClose();
-                delete m_rxInPa;
-                delete m_rxOutPa;
-                if(two_tx)
-                    delete m_txOutPa;
-                destroy_fifos();
-                destroy_src();
-                deleteEQFilters(g_rxUserdata);
-                delete g_rxUserdata;
-                m_RxRunning = false;
-                Pa_Terminate();
-                return;
+        // Set sound card callbacks
+        auto errorCallback = [&](IAudioDevice&, std::string error, void*)
+        {
+            CallAfter([&]() {
+                wxMessageBox(wxString::Format("Error encountered while processing audio: %s", error), wxT("Error"), wxOK);
+            });
+        };
```

Comment:
I'm not across this C++ syntax to understand this fully - but is this popping up a Dialog in the thread/context of a callback?  If so, I'm wondering it that's OK?  It would be a good idea to exercise any code that generates a MessageBox.

Reply:
Yep, CallAfter() is provided by wxWidgets to ensure that all GUI related actions happen on the GUI thread. Bad things happen on the wxWidgets side otherwise.

Reply:
BTW the [&] ... syntax is a [C++11 lambda](https://en.cppreference.com/w/cpp/language/lambda).